crypto/rand/drbg_lib.c

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright 2011-2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/rand.h>
+#include "rand_lcl.h"
+#include "rand_drbg_lcl.h"
+
+/*
+ * Support framework for SP800-90 DRBGs
+ */
+
+
+int RAND_DRBG_set(DRBG_CTX *dctx, int type, unsigned int flags)
+{
+    int ret = 1;
+
+    dctx->status = DRBG_STATUS_UNINITIALISED;
+    dctx->xflags = flags;
+    dctx->type = type;
+    dctx->iflags = 0;
+    dctx->entropy_blocklen = 0;
+
+    switch (type) {
+    default:
+        RANDerr(RAND_F_RAND_DRBG_SET, RAND_R_UNSUPPORTED_DRBG_TYPE);
+        return -2;
+    case 0:
+        /* Uninitialized; that's okay. */
+        return 1;
+#if 0
+    case NID_sha224:
+    case NID_sha256:
+    case NID_sha384:
+    case NID_sha512:
+        ret = drbg_hash_init(dctx);
+        break;
+    case NID_hmacWithSHA1:
+    case NID_hmacWithSHA224:
+    case NID_hmacWithSHA256:
+    case NID_hmacWithSHA512:
+        ret = drbg_hmac_init(dctx);
+        break;
+#endif
+    case NID_aes_128_ctr:
+    case NID_aes_192_ctr:
+    case NID_aes_256_ctr:
+        ret = ctr_init(dctx);
+        break;
+    }
+
+    if (ret < 0)
+        RANDerr(RAND_F_RAND_DRBG_SET, RAND_R_ERROR_INITIALISING_DRBG);
+    return ret;
+}
+
+DRBG_CTX *RAND_DRBG_new(int type, unsigned int flags)
+{
+    DRBG_CTX *dctx = OPENSSL_zalloc(sizeof(*dctx));
+
+    if (dctx == NULL) {
+        RANDerr(RAND_F_RAND_DRBG_NEW, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    if (RAND_DRBG_set(dctx, type, flags) < 0) {
+        OPENSSL_free(dctx);
+        return NULL;
+    }
+    return dctx;
+}
+
+void RAND_DRBG_free(DRBG_CTX *dctx)
+{
+    if (dctx == NULL)
+        return;
+    if (dctx->uninstantiate != NULL)
+        dctx->uninstantiate(dctx);
+
+    CRYPTO_free_ex_data(CRYPTO_EX_INDEX_DRBG, dctx, &dctx->ex_data);
+
+    /* Don't free up default DRBG */
+    if (dctx == RAND_DRBG_get_default()) {
+        memset(dctx, 0, sizeof(DRBG_CTX));
+        dctx->type = 0;
+        dctx->status = DRBG_STATUS_UNINITIALISED;
+    } else {
+        OPENSSL_cleanse(&dctx->ctr, sizeof(dctx->ctr));
+        OPENSSL_free(dctx);
+    }
+}
+
+static size_t get_entropy(DRBG_CTX *dctx, unsigned char **pout,
```

Comment:
Comparing the the original FIPS code for fips_get_entropy() and your get_entropy() it appears to me that [rt.openssl.org #4055] (Github Issue #2443) hasn't been addressed yet, see my outdated review and https://mta.openssl.org/pipermail/openssl-dev/2015-October/002950.html.

Will this issue be fixed in this PR, or will be there a separate PR?